File: backend/src/controllers/patient.controller.ts
import { Request, Response, NextFunction } from 'express';
import * as patientService from '../services/patient.service';

export const createPatient = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const result = await patientService.createPatient(req.body);
    res.status(201).json({
      success: true,
      message: 'Patient registered successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPatient = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await patientService.getPatientById(id);
    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPatientByUserId = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = parseInt(req.params.userId, 10);
    const result = await patientService.getPatientByUserId(userId);
    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getAllPatients = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);
    const search = req.query.search as string;
    const sortBy = req.query.sortBy as string || 'createdAt';

    const result = await patientService.getAllPatients(page, size, search, sortBy);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const updatePatient = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await patientService.updatePatient(id, req.body);
    res.status(200).json({
      success: true,
      message: 'Patient updated successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const addMedicalHistory = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await patientService.addMedicalHistory(id, req.body);
    res.status(201).json({
      success: true,
      message: 'Medical history added successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const addEmergencyContact = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await patientService.addEmergencyContact(id, req.body);
    res.status(201).json({
      success: true,
      message: 'Emergency contact added successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

File: backend/patient-service/src/main/java/com/hospital/patient/controller/PatientController.java
package com.hospital.patient.controller;

import com.hospital.common.dto.ApiResponse;
import com.hospital.common.dto.PagedResponse;
import com.hospital.patient.dto.*;
import com.hospital.patient.service.PatientService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/patients")
@RequiredArgsConstructor
@Tag(name = "Patients", description = "Patient management operations")
public class PatientController {

    private final PatientService patientService;

    @PostMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Register Patient", description = "Create a new patient profile")
    public ResponseEntity<ApiResponse<PatientResponse>> createPatient(
            @Valid @RequestBody PatientCreateRequest request) {
        PatientResponse patient = patientService.createPatient(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Patient registered", patient));
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Get Patient", description = "Get patient details by ID")
    public ResponseEntity<ApiResponse<PatientResponse>> getPatient(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.success(patientService.getPatientById(id)));
    }

    @GetMapping("/user/{userId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Get Patient by User ID", description = "Get patient profile by user ID")
    public ResponseEntity<ApiResponse<PatientResponse>> getPatientByUserId(@PathVariable Long userId) {
        return ResponseEntity.ok(ApiResponse.success(patientService.getPatientByUserId(userId)));
    }

    @GetMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'RECEPTIONIST')")
    @Operation(summary = "List Patients", description = "Get paginated list of patients")
    public ResponseEntity<ApiResponse<PagedResponse<PatientResponse>>> getAllPatients(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "createdAt") String sortBy) {
        return ResponseEntity.ok(ApiResponse.success(
                patientService.getAllPatients(page, size, search, sortBy)));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Update Patient", description = "Update patient information")
    public ResponseEntity<ApiResponse<PatientResponse>> updatePatient(
            @PathVariable Long id, @Valid @RequestBody PatientCreateRequest request) {
        return ResponseEntity.ok(ApiResponse.success("Patient updated",
                patientService.updatePatient(id, request)));
    }

    @PostMapping("/{id}/medical-history")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR')")
    @Operation(summary = "Add Medical History", description = "Add medical history record")
    public ResponseEntity<ApiResponse<MedicalHistoryDto>> addMedicalHistory(
            @PathVariable Long id, @Valid @RequestBody MedicalHistoryDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Medical history added",
                        patientService.addMedicalHistory(id, dto)));
    }

    @PostMapping("/{id}/emergency-contacts")
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Add Emergency Contact", description = "Add emergency contact")
    public ResponseEntity<ApiResponse<EmergencyContactDto>> addEmergencyContact(
            @PathVariable Long id, @Valid @RequestBody EmergencyContactDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Emergency contact added",
                        patientService.addEmergencyContact(id, dto)));
    }
}
